File: translate-batch.py
#!/usr/bin/env python3
"""Batch translate quiz questions + options to Slovak."""

import json
import os
import sys
import time
import urllib.request
import urllib.error
import ssl

# Disable SSL verification (macOS Python issue)
ctx = ssl.create_default_context()
ctx.check_hostname = False
ctx.verify_mode = ssl.CERT_NONE

SB_URL = os.environ["SUPABASE_URL"]
SB_KEY = os.environ["SUPABASE_SERVICE_ROLE_KEY"]
OAI_KEY = os.environ["OPENAI_API_KEY"]

SB_HEADERS = {
    "apikey": SB_KEY,
    "Authorization": f"Bearer {SB_KEY}",
    "Content-Type": "application/json",
}

def sb_get(path):
    url = f"{SB_URL}/rest/v1/{path}"
    req = urllib.request.Request(url, headers=SB_HEADERS)
    with urllib.request.urlopen(req, context=ctx) as resp:
        return json.loads(resp.read())

def sb_patch(table, row_id, data):
    url = f"{SB_URL}/rest/v1/{table}?id=eq.{row_id}"
    payload = json.dumps(data).encode()
    req = urllib.request.Request(url, data=payload, headers=SB_HEADERS, method="PATCH")
    try:
        urllib.request.urlopen(req, context=ctx)
    except urllib.error.HTTPError as e:
        print(f"  PATCH error {table} id={row_id}: {e.code}")

def translate(texts):
    prompt = (
        "Translate these English texts to Slovak (slovenčina). "
        "NEVER use Czech words. Return JSON: {\"translations\": [\"...\", ...]}\n\n"
        f"Texts:\n{json.dumps(texts, ensure_ascii=False)}"
    )
    body = json.dumps({
        "model": "gpt-4o-mini",
        "messages": [{"role": "user", "content": prompt}],
        "temperature": 0.2,
        "max_tokens": 4000,
        "response_format": {"type": "json_object"},
    }).encode()

    req = urllib.request.Request(
        "https://api.openai.com/v1/chat/completions",
        data=body,
        headers={
            "Authorization": f"Bearer {OAI_KEY}",
            "Content-Type": "application/json",
        },
    )
    with urllib.request.urlopen(req, context=ctx) as resp:
        data = json.loads(resp.read())

    content = data["choices"][0]["message"]["content"]
    parsed = json.loads(content)

    if isinstance(parsed, list):
        return parsed
    if "translations" in parsed:
        return parsed["translations"]
    vals = list(parsed.values())
    if vals and isinstance(vals[0], list):
        return vals[0]
    return vals


def translate_questions():
    print("=== Translating Questions ===\n")
    total = 0

    while True:
        rows = sb_get("cb_quiz_questions?select=id,question_text&question_text_sk=is.null&order=id&limit=20")
        if not rows:
            break

        texts = [r["question_text"] for r in rows]
        try:
            translations = translate(texts)
        except Exception as e:
            print(f"  GPT error: {e}")
            break

        for i, row in enumerate(rows):
            if i >= len(translations):
                break
            sb_patch("cb_quiz_questions", row["id"], {"question_text_sk": translations[i]})

        total += len(rows)
        print(f"  Questions translated: {total}")
        time.sleep(0.3)

    print(f"\nDone: {total} questions.\n")


def translate_options():
    print("=== Translating Options ===\n")
    total = 0

    while True:
        rows = sb_get("cb_quiz_options?select=id,option_text&option_text_sk=is.null&order=id&limit=40")
        if not rows:
            break

        texts = [r["option_text"] for r in rows]
        try:
            translations = translate(texts)
        except Exception as e:
            print(f"  GPT error: {e}")
            break

        for i, row in enumerate(rows):
            if i >= len(translations):
                break
            sb_patch("cb_quiz_options", row["id"], {"option_text_sk": translations[i]})

        total += len(rows)
        print(f"  Options translated: {total}")
        time.sleep(0.3)

    print(f"\nDone: {total} options.\n")


if __name__ == "__main__":
    print("Batch SK Translation\n")
    translate_questions()
    translate_options()
    print("All translations complete!")

File: translate-batch.ts
import 'dotenv/config';
import { createClient } from '@supabase/supabase-js';

const sb = createClient(process.env.SUPABASE_URL!, process.env.SUPABASE_SERVICE_ROLE_KEY!);
const OPENAI_KEY = process.env.OPENAI_API_KEY!;

const BATCH_SIZE = 20; // questions per GPT call
const DELAY = 500; // ms between API calls

async function translate(texts: string[]): Promise<string[]> {
  const prompt = `Translate these English texts to Slovak (slovenčina). NEVER use Czech. Return a JSON array of translated strings in the same order.

Texts:
${JSON.stringify(texts)}`;

  const res = await fetch('https://api.openai.com/v1/chat/completions', {
    method: 'POST',
    headers: { 'Authorization': `Bearer ${OPENAI_KEY}`, 'Content-Type': 'application/json' },
    body: JSON.stringify({
      model: 'gpt-4o-mini',
      messages: [{ role: 'user', content: prompt }],
      temperature: 0.2,
      max_tokens: 4000,
      response_format: { type: 'json_object' },
    }),
  });

  const data = await res.json();
  const content = data.choices?.[0]?.message?.content || '{}';
  const parsed = JSON.parse(content);
  // Handle both array and object with array property
  if (Array.isArray(parsed)) return parsed;
  const vals = Object.values(parsed);
  if (vals.length === 1 && Array.isArray(vals[0])) return vals[0] as string[];
  return vals as string[];
}

async function sleep(ms: number) {
  return new Promise(r => setTimeout(r, ms));
}

async function translateQuestions() {
  console.log('=== Translating Questions ===\n');

  let offset = 0;
  let totalDone = 0;

  while (true) {
    const { data: questions, error } = await sb
      .from('cb_quiz_questions')
      .select('id, question_text')
      .is('question_text_sk', null)
      .order('id')
      .range(offset, offset + BATCH_SIZE - 1);

    if (error) { console.error('Fetch error:', error); break; }
    if (!questions || questions.length === 0) break;

    const texts = questions.map(q => q.question_text);

    try {
      const translations = await translate(texts);

      for (let i = 0; i < questions.length; i++) {
        const sk = translations[i];
        if (!sk) { console.warn(`  Skip q=${questions[i].id} — no translation`); continue; }

        const { error: upErr } = await sb
          .from('cb_quiz_questions')
          .update({ question_text_sk: sk })
          .eq('id', questions[i].id);

        if (upErr) console.error(`  Error updating q=${questions[i].id}:`, upErr);
      }

      totalDone += questions.length;
      console.log(`  Translated ${totalDone} questions (batch of ${questions.length})`);
    } catch (err) {
      console.error('  Translation error:', err);
      offset += BATCH_SIZE; // skip batch on error
    }

    await sleep(DELAY);
    // Don't increment offset — we query for null, so translated ones won't appear again
  }

  console.log(`\nDone: ${totalDone} questions translated.\n`);
}

async function translateOptions() {
  console.log('=== Translating Options ===\n');

  let totalDone = 0;

  while (true) {
    const { data: options, error } = await sb
      .from('cb_quiz_options')
      .select('id, option_text')
      .is('option_text_sk', null)
      .order('id')
      .range(0, BATCH_SIZE * 2 - 1); // larger batches for short texts

    if (error) { console.error('Fetch error:', error); break; }
    if (!options || options.length === 0) break;

    const texts = options.map(o => o.option_text);

    try {
      const translations = await translate(texts);

      for (let i = 0; i < options.length; i++) {
        const sk = translations[i];
        if (!sk) { console.warn(`  Skip opt=${options[i].id} — no translation`); continue; }

        const { error: upErr } = await sb
          .from('cb_quiz_options')
          .update({ option_text_sk: sk })
          .eq('id', options[i].id);

        if (upErr) console.error(`  Error updating opt=${options[i].id}:`, upErr);
      }

      totalDone += options.length;
      console.log(`  Translated ${totalDone} options (batch of ${options.length})`);
    } catch (err) {
      console.error('  Translation error:', err);
      // fetch fresh batch next iteration
    }

    await sleep(DELAY);
  }

  console.log(`\nDone: ${totalDone} options translated.\n`);
}

async function main() {
  console.log('🌍 Batch SK Translation\n');
  await translateQuestions();
  await translateOptions();
  console.log('✅ All translations complete!');
}

main().catch(err => { console.error('Fatal:', err); process.exit(1); });
